_posts/07-03-01-Databases_PDO.md

```diff
@@ -1,77 +1,64 @@
+Esto es posible gracias al uso de las declaraciones (statements) y parámetros vinculados (bound parameters) de PDO.
+
+Supongamos que un script PHP recibe un ID numérico como un parámetro de una consulta. Este ID se usará para buscar un registro de usuario en una base de datos. Esta es la manera `incorrecta` de hacer esto:
 
 {% highlight php %}
-<?php
 $pdo = new PDO('sqlite:/path/db/users.db');
-$pdo->query("SELECT name FROM users WHERE id = " . $_GET['id']); // <-- NO!
+$pdo->query("SELECT nombre FROM usuarios WHERE id = " . $_GET['id']); // <-- NO!
 {% endhighlight %}
 
-This is terrible code. You are inserting a raw query parameter into a SQL query. This will get you hacked in a
-heartbeat, using a practice called [SQL Injection]. Just imagine if a hacker passes in an inventive `id` parameter by
-calling a URL like `http://domain.com/?id=1%3BDELETE+FROM+users`. This will set the `$_GET['id']` variable to `1;DELETE
-FROM users` which will delete all of your users! Instead, you should sanitize the ID input using PDO bound parameters.
+Ese es un código terrible. Al hacerlo, estaría insertando un parámetro _crudo_ en su sentencia SQL. Hacerlo significaría que podrían hackearle más rápido que inmediatamente, usando una técnica llamada [SQL Injection](http://wiki.hashphp.org/Validation). Sólo imagine que un hacker pase un parámetro inventado `id` llamando a una URL como `http://dominio.com/?id=1%3BDELETE+FROM+usuarios`. Esto le asignaría el valor `1;DELETE FROM users` a la variable global `$_GET['id']` ¡lo cual borraría todos los registros de su tabla `usuarios`! Para evitarlo, debe _desinfectar_ la entrada ID usando parámetros vinculados (bound parameters).
 
 {% highlight php %}
-<?php
 $pdo = new PDO('sqlite:/path/db/users.db');
-$stmt = $pdo->prepare('SELECT name FROM users WHERE id = :id');
-$id = filter_input(INPUT_GET, 'id', FILTER_SANITIZE_NUMBER_INT); // <-- filter your data first (see [Data Filtering](#data_filtering)), especially important for INSERT, UPDATE, etc.
-$stmt->bindParam(':id', $id, PDO::PARAM_INT); // <-- Automatically sanitized for SQL by PDO
+$stmt = $pdo->prepare('SELECT nombre FROM usuarios WHERE id = :id');
+$stmt->bindParam(':id', $_GET['id'], PDO::PARAM_INT); // <-- Desinfectado automáticamente por PDO
 $stmt->execute();
 {% endhighlight %}
 
-This is correct code. It uses a bound parameter on a PDO statement. This escapes the foreign input ID before it is
-introduced to the database preventing potential SQL injection attacks.
-
-For writes, such as INSERT or UPDATE, it's especially critical to still [filter your data](#data_filtering) first and sanitize it for other things (removal of HTML tags, JavaScript, etc).  PDO will only sanitize it for SQL, not for your application.
+Ese es el código correcto. Se usa un parámetro vinculado (bound parameter) en una declaración PDO. Esto escapa la entrada externa ID antes de que esta sea introducida a la base de datos, previendo posibles ataques de SQL Injection.
```

Comment:
Creo que falto traducir el seguno parrafo eliminado "For writes, such as INSERT or UPDATE, it's especially..."